enhancements/machine-api/cluster-autoscaler-integration.md

```diff
@@ -0,0 +1,456 @@
+Cluster Autoscaler with information about the resources within a cluster and the resources available for scaling.
+
+MachineSets and MachineDeployments will be opted-in to autoscaling if they are annotated with two annotations, which
+denote the maximum and minimum sizes that the autoscaler should scale the Node group within, eg:
+```yaml
+metadata:
+  annotations:
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-max-size": "6"
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-min-size": "1"
+```
+
+### User Stories
+
+#### Story 1
+
+As a user of OpenShift, I would like to be able to autoscale workloads on my cluster as required by my users, and have
+the cluster automatically expand to meet the new capacity requirements
+
+#### Story 2
+
+As a user of OpenShift, if the workloads on my cluster reduce leaving unused capcity, I would like my cluster to
+automatically scale back to reduce cost and wasted resources.
+
+#### Story 3
+
+As a developer of OpenShift, I would like an agnostic integration that specifies a minimal API contract for autoscaling
+so that it is compatible with multiple versions and implementations of the machine API simultaneously to reduce the
+burden of fork/extend style development.
+
+### Implementation Details
+
+This integration will satisfy the [CloudProvider](https://github.com/kubernetes/autoscaler/blob/a00bf59159bda6d70e86741764802d11e7d253f4/cluster-autoscaler/cloudprovider/cloud_provider.go#L48-L91)
+interface by mapping the [NodeGroup](https://github.com/kubernetes/autoscaler/blob/a00bf59159bda6d70e86741764802d11e7d253f4/cluster-autoscaler/cloudprovider/cloud_provider.go#L103-L170)
+concept to MachineSets and MachineDeployments that opt-in to autoscaling.
+
+#### Opting in to Autoscaling
+
+In order for a MachineSet or MachineDeployment to be considered for autoscaling, the autoscaler integration will need
+autoscaling bounds for the NodeGroup, ie. it will need to know the maximum and minimum size that the MachineSet/
+MachineDeployment should be.
+
+This information will be provided to the Cluster Autoscaler by way of annotations on the MachineSet/MachineDeployment, eg:
+```yaml
+metadata:
+  annotations:
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-max-size": "6"
+    - "machine.openshift.io/cluster-api-autoscaler-node-group-min-size": "1"
+```
+
+#### Provider
+
+The Provider will implement the [CloudProvider](https://github.com/kubernetes/autoscaler/blob/a00bf59159bda6d70e86741764802d11e7d253f4/cluster-autoscaler/cloudprovider/cloud_provider.go#L48-L91)
+interface that is required to integrate Machine API with the cluster autoscaler.
+
+It will essentially wrap the Machine Controller which is responsible for maintaining the state of the cluster from an
+autoscaling perspective.
+
+#### Machine Controller
+
+The Machine Controller will be responsible for coordinating the underlying MachineSets and MachineDeployments to provide
```

Comment:
nit: calling it a "Machine Controller" to me implies it is going to watch Machines as a primary resource. But my impression is that it probably won't need to watch Machines at all.

Reply:
I agree, not a great name. It actually does a lot of things, it's kind of the client for the cloud provider in a way, providing access to all the required resources for making decisions and modifying the state of the cluster

Since this is a retrospective enhancement proposal, I won't update this doc for now, but will raise the issue with the team and see if we can come up with a better name for it

Reply:
may be include a link https://github.com/openshift/kubernetes-autoscaler/blob/4abdca547be45a251ba33486324f0cac8664ca25/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go#L51-L67